### ISSUE TITLE
Is it neccessary to install intel_sgx_psw packege when using Grapehen

### ISSUE THREAD

**Issue by kleirama** [oscarlab/graphene#1558]
Hi,

I have read through Documentation directory here in Graphene repository and It seems that intel sgx psw is not mentioned as a requirement. Do we need to install it while using Graphene-SGX?

I am not using https://github.com/intel/linux-sgx-driver  nor https://github.com/intel/linux-sgx to install Intel SGX driver and SDK and PSW. I think that when you mention that you need to go through the above links to install Intel SGX SDK and Intel SGX Driver you implicitly may have implied Intel SGX PSW since Intel SGX PSW and SDK sit in the same repository [](https://github.com/intel/linux-sgx)

In my case I am downloading binary files from https://download.01.org/intel-sgx/sgx-linux/2.9.1/distro/ubuntu18.04-server/ and then I use this guide to install the binary files  https://download.01.org/intel-sgx/linux-1.6/docs/Intel_SGX_SDK_Installation_Guide_Linux_1.6_Open_Source.pdf. Do I need to download also PSW binary files or do I just need to download aesm service and SGX SDK and SGX driver files?

Cheers,

**Comment by dimakuv** [comment]
Yes, it is necessary to install Intel SGX PSW as well. In particular, PSW installs the Launch Enclave, the Quoting Enclave (these are the so-called "ae" or "Architectural Enclaves"), and the AESM service -- all of which are needed by Graphene. Also, if you're using the DCAP software infrastructure on your system, you need to install all required DCAP libraries -- Graphene needs `libsgx-dcap-ql` and all its dependencies.

Simply put, I would recommend to install *all* those packages mentioned in the Installation Guide you linked.

If you're specifically looking for a *minimum* set of packages to install, unfortunately, we haven't done this exercise yet. It is unclear what are the *absolute minimum* requirements of Graphene regarding the SGX SDK/PSW/DCAP software...

**Comment by mkow** [comment]
The question was answered and no follow-up from the reporter for quite some time, closing.

**Comment by dimakuv** [comment]
Since I'm looking at this issue now, I'll just add some more technical details.

We still haven't done the exercise of determining the minimum set of Intel SGX SDK/PSW packages for Graphene-SGX to work. However, I can give some context on what Graphene-SGX uses from SDK/PSW:

1. Graphene-SGX uses the Launch Enclave to retrieve the EINITTOKEN **if** the SGX-enabled system is not DCAP-based (more specifically, if the system does **not** have Flexible Launch Control hardware feature *and* the system uses the legacy non-DCAP SGX driver). To communicate with the Launch Enclave, Graphene-SGX relies on the AESM daemon with the corresponding plugin. The AESM daemon is part of the Intel SGX PSW. Thus, if the system is not DCAP-based, Graphene-SGX needs Intel SGX PSW to be installed on the system. To put it differently, if the system **is** DCAP-based, Graphene-SGX doesn't need Intel SGX PSW (Graphene-SGX still creates a token file, but this file is a dummy).

2. Graphene-SGX uses the Quoting Enclave to retrieve the SGX quote **in all cases when SGX remote attestation is used**. If your particular use case of Graphene-SGX doesn't need SGX remote attestation (cases of testing, debugging, etc.), then Graphene-SGX doesn't need Intel SGX PSW installed.

These are the only two items I can think of right now. Please note that Graphene-SGX is largely a **standalone** framework. It does **not** reuse the Intel SGX SDK functionality like "Common Enclave Loader" or the "Trusted Standard C Library" or the "Protected File System" (the latter is ported from Intel SGX SDK as sources, but *not* as a shared-library dependency).

So, on a DCAP-based system with a newer SGX driver and that doesn't use SGX remote attestation, it should be possible to build and run Graphene-SGX without installing Intel SGX SDK/PSW. Hopefully this theoretical exercise is true in reality...
